module/zfs/dmu_send.c

```diff
@@ -2484,18 +2484,89 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
 		    (drro->drr_blksz != doi.doi_data_block_size ||
 		    nblkptr < doi.doi_nblkptr ||
 		    indblksz != doi.doi_metadata_block_size ||
-		    drro->drr_nlevels < doi.doi_indirection)) {
+		    drro->drr_nlevels < doi.doi_indirection ||
+		    drro->drr_dn_slots != doi.doi_dnodesize >> DNODE_SHIFT)) {
 			err = dmu_free_long_range_raw(rwa->os,
 			    drro->drr_object, 0, DMU_OBJECT_END);
 			if (err != 0)
 				return (SET_ERROR(EINVAL));
 		} else if (drro->drr_blksz != doi.doi_data_block_size ||
-		    nblkptr < doi.doi_nblkptr) {
+		    nblkptr < doi.doi_nblkptr ||
+		    drro->drr_dn_slots != doi.doi_dnodesize >> DNODE_SHIFT) {
 			err = dmu_free_long_range(rwa->os, drro->drr_object,
 			    0, DMU_OBJECT_END);
 			if (err != 0)
 				return (SET_ERROR(EINVAL));
 		}
+
```

Comment:
This section might have code the other platforms wants, the rest appear to be regarding large dnodes

Reply:
Yeah.... its a bit hard to show the split here. Basically other platforms need the first `if` and the `else`. The `else if (err == EEXIST)` is for large dnodes, along with `if (drro->drr_dn_slots > 1)`, and the `EEXIST` check on line 3274.

Reply:
Ok, ends up looking a bit like
https://github.com/openzfsonosx/zfs/commit/9c429ad944d88588ffceb300b4f0ef80ddf85b08